src/plugindata.h

```diff
@@ -263,8 +265,66 @@ void plugin_configure_single(GtkWidget *parent);
 void plugin_help(void);
 void plugin_cleanup(void);
 
+/** Called by Geany when a plugin library is loaded.
+ *
+ * This is the original entry point. Implement and export this function to be loadable at all.
+ * Then fill in GeanyPlugin::info and GeanyPlugin::funcs of the passed @p plugin. Finally
+ * GEANY_PLUGIN_REGISTER() and specify a minimum supported API version.
+ *
+ * For all glory details please read @ref howto.
+ *
+ * @param plugin The unique plugin handle to your plugin. You must set some fields here.
+ * @param module The GModule associated with your plugin library.
+ * @param geany_api_version The API version exposed by the Geany instance that loads your plugin.
+ *           This probably not the same as the one your plugin was compiled against.
+ *
+ * @since 1.26 (API 225)
+ * @see @ref howto
+ */
+void geany_load_module(GeanyPlugin *plugin, GModule *module, gint geany_api_version);
+
 #endif
 
+/** Callback functions that need to be implemented for every plugin.
+ *
+ * These callbacks should be registered by the plugin within Geany's call to
+ * geany_load_module() by calling geany_plugin_register() with an instance of this type.
+ *
+ * Geany will then call the callbacks at appropriate times. Each gets passed the
+ * plugin-defined data pointer as well as the corresponding GeanyPlugin instance
+ * pointer.
+ *
+ * @since 1.26 (API 225)
+ * @see @ref howto
+ **/
+struct GeanyPluginFuncs
+{
+	/** Array of plugin-provided signal handlers @see PluginCallback */
+	PluginCallback *callbacks;
+	/** Called to initialize the plugin, when the user activates it (must not be @c NULL) */
+	gboolean    (*init)      (GeanyPlugin *plugin, gpointer pdata);
+	/** plugins configure dialog, optional (can be @c NULL) */
+	GtkWidget*  (*configure) (GeanyPlugin *plugin, GtkDialog *dialog, gpointer pdata);
+	/** Called when the plugin should show some help, optional (can be @c NULL) */
+	void        (*help)      (GeanyPlugin *plugin, gpointer pdata);
+	/** Called when the plugin is disabled or when Geany exits (must not be @c NULL) */
```

Comment:
I meant it in the general meaning:

> re·gres·sion: 1. a return to a former or less developed state.

Rather than being improved, it's actually worse than before. I didn't mean it like a "software regression" where it broke something that once worked (though it kind of applies still).

In any case, as long as `free_func` is called exactly once, only when the plugin is de-activated, and it can be set when the plugin is activated, then it's fine. The only weird thing left is allowing a plugin to call `geany_plugin_set_data()` from `geany_load_module()`, it ought to guard against that as it's likely to cause confusion/bugs (or at least it should be documented to not be recommended or something).